### PR TITLE
Add Windows OneCore targets.

### PR DESCRIPTION
**Pull request by dot-asm**

### CODE REVIEW

**Comment by dot-asm** [comment]
Richard, it's mostly about middle commit...

File: Configurations/10-main.conf [hunk 1]
```diff
@@ -1418,8 +1420,10 @@ my %targets = (
         multilib         => "-x64",
     },
     "VC-WIN32" => {
-        # x86 Win32 target defaults to ANSI API, if you want UNICODE,
-        # configure with 'perl Configure VC-WIN32 -DUNICODE -D_UNICODE'
+        # x86 Win32 target used to default to ANSI API, but not anymore,
+        # UNICODE is the only option. The original reason for providing
+        # UNICODE as opt-in option was becase ANSI API was *native*
```
Comment (levitte) [inline Configurations/10-main.conf]: 'becase' => 'because'

File: Configurations/50-win-arm.conf [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        lflags          => add("/NODEFAULTLIB:kernel32.lib"),
+        ex_libs         => "onecore.lib onecore_downlevel.lib",
+    },
+    "VC-WIN64-ARM" => {
+        inherit_from    => [ "VC-noCE-common" ],
+        defines         => add("_ARM_WINAPI_PARTITION_DESKTOP_SDK_AVAILABLE"),
+        bn_ops          => "SIXTY_FOUR_BIT RC4_CHAR EXPORT_VAR_AS_FN",
+        lflags          => add("/NODEFAULTLIB:kernel32.lib"),
+        ex_libs         => "onecore.lib onecore_downlevel.lib",
+    },
```
Comment (levitte) [inline Configurations/50-win-arm.conf]: So, errr, where is `VC-WIN32-ARM` supposed to be installed?  I notice that you're not adding any flavor, so it will end up being installed like a VC-W64 flavor...

File: Configurations/10-main.conf [hunk 3]
```diff
@@ -1343,6 +1345,7 @@ my %targets = (
         thread_scheme    => "winthreads",
         dso_scheme       => "win32",
         apps_aux_src     => add("win32_init.c"),
+        build_scheme     => add("VC-common", { separator => undef }),
```
Comment (levitte) [inline Configurations/10-main.conf]: Since this affects installation directories, I'm not at all sure about this.  I much prefer to have explicit flavors (for example `VC-W64` that you removed) than "trying to be smart".

File: Configurations/50-win-arm.conf (hunk 2 above)
Comment (dot-asm) [reply Configurations/50-win-arm.conf]: Well, define "installation" in cross-compilation scenario. In *pure* cross-compilation scenario when target file system has *nothing* to do with host's. Nor has it any of host's facilities. [E.g. what meaning do headers and lib files have to target system when it doesn't and never will have compiler?] Easiest to imagine on CE's example. There is no C: on CE device, or `Program Files`. So why one would compile host's values into target's binary code? Answer is that you pretty much have no other choice but to specify **target's** path explicitly with --prefix [and the other one], then specify DESTDIR at nmake install time at host, and finally do whatever undefined magic it takes to transfer executable code to the target system. [Note that you might as well choose to skip nmake install.] But let's consider that you didn't specify --prefix in this win-arm case. There actually is `C:\Program Files` [at least] on my device. So it's actually not inappropriate **default** for this target installation. In fact, if you think about it, it wouldn't be inappropriate choice for any non-dual-ABI non-CE system. "Non-dual-ABI" means that you can execute only one kind of binaries. In other words having "VC-W64" install path is actually meaningful default.

As for "VC-W64" flavour in more general sense. With last sentence in mind one can argue that it's more appropriate to actually view it as "VC-non-dual-ABI" or ... "VC-common" ;-) Also note that all that "VC-W32" dance is rather about dual-ABI than anything else. So imagine we rename it to "VC-WOW" instead. Wouldn't it be helpful? Just in case "WOW" stands for "Windows on Windows" and is kind of established term.

[Just in case for reference. Is it hard to get hold of Windows ARM device now? Not really, there are images for Raspberry Pi. Though I myself don't run it on Raspberry Pi.]

File: Configurations/10-main.conf (hunk 3 above)
Comment (dot-asm) [reply Configurations/10-main.conf]: See above.

File: Configurations/50-win-arm.conf (hunk 2 above)
Comment (dot-asm) [reply Configurations/50-win-arm.conf]: It's not inappropriate to add remark about host vs. target file systems into commentary section, so I'll do that...

File: Configurations/50-win-arm.conf [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+my %targets = (
+    "VC-WIN32-ARM" => {
+        inherit_from    => [ "VC-noCE-common" ],
+        defines         => add("_ARM_WINAPI_PARTITION_DESKTOP_SDK_AVAILABLE"),
+        bn_ops          => "BN_LLONG RC4_CHAR EXPORT_VAR_AS_FN",
+        # /NODEFAULTLIB:kernel32.lib is needed, because MSVCRT.LIB has
+        # direct reference to kernel32.lib, but we don't actually want
+        # it in "onecore" build.
+        lflags          => add("/NODEFAULTLIB:kernel32.lib"),
+        ex_libs         => "onecore.lib onecore_downlevel.lib",
```
Comment (levitte) [inline Configurations/50-win-arm.conf]: Will onecode_downlevel.lib still be needed when #4965 gets merged?

Comment (dot-asm) [reply Configurations/50-win-arm.conf]: It would be possible to omit it [and add additional define].

Comment (dot-asm) [reply Configurations/50-win-arm.conf]: ... but what's "needed" anyway? I can confirm that binaries built as suggested do run on a **certified** Windows IoT device, even though they are linked with onecore_downlevel.lib. [Well, "certified" means that I'm told that it is, not that I certified it myself.] One can argue that doing so make it non-compliant with some "big picture", but you know what? Said "big picture" is not really self-consistent. They say "onecore only, here is the list", but you also end-up with reference to CRT. In other words it's not really that categorical as "onecore only". It might happen that reliance on CRT disqualifies binary code as compliant with "big picture", but then we are out of luck anyway... [In the context you might recall that at some point I actually suggested to consider removal or rather minimization of the reliance on MS CRT, but met no support for the idea.]

File: Configurations/50-win-arm.conf (hunk 2 above)
Comment (dot-asm) [reply Configurations/50-win-arm.conf]: > rename it [VC-W32] to "VC-WOW"

Shall we?

**Comment by dot-asm** [comment]
On additional note it effectively supersedes #4920.

File: Configurations/windows-makefile.tmpl [hunk 5]
```diff
@@ -14,18 +14,18 @@
 
  (our $sover_dirname = $config{shlib_version_number}) =~ s|\.|_|g;
 
+ my $build_scheme = $target{build_scheme};
+ my $build_flavour = $build_scheme->[$#$build_scheme];
```
Comment (dot-asm) [inline Configurations/windows-makefile.tmpl]: In the context of "VC-W32" vs. "VC-WOW" it might be appropriate to name this variable as `$install_flavour`.

Comment (levitte) [reply Configurations/windows-makefile.tmpl]: Who says that the value is limited to installation stuff? The point was for the value to indicate a categorisation, to be used as we see fit.

File: Configurations/50-win-arm.conf (hunk 2 above)
Comment (levitte) [reply Configurations/50-win-arm.conf]: I see no problem with VC-WOW per se.

File: Configurations/windows-makefile.tmpl (hunk 5 above)
Comment (dot-asm) [reply Configurations/windows-makefile.tmpl]: > Who says?

I do :-)

> categorisation

It still categorizes things :-) But on serious note, this is used [to pull environment variables that is then used] as pre-compiled path reflecting where thins will be installed, and of course as destination directory for nmake install. In both cases there is notion of "install" involved, so I'd say it's not inappropriate name...

**Comment by levitte** [comment]
Please rebase?

**Comment by dot-asm** [comment]
Re-based. I've also added x86 onecore targets. As discussed in #4965, suggestion is to solve omission of onecore_downlevel.lib as separate problem.

File: Configurations/10-main.conf [hunk 6]
```diff
@@ -1449,8 +1452,10 @@ my %targets = (
         multilib         => "-x64",
     },
     "VC-WIN32" => {
-        # x86 Win32 target defaults to ANSI API, if you want UNICODE,
-        # configure with 'perl Configure VC-WIN32 -DUNICODE -D_UNICODE'
+        # x86 Win32 target used to default to ANSI API, but not anymore,
+        # UNICODE is the only option. The original reason for providing
+        # UNICODE as opt-in option was because ANSI API was *native*
+        # system interface for [no longer supported] Windows 9x.
```
Comment (levitte) [inline Configurations/10-main.conf]: This comment should be moved to `VC-noCE-common`, where `UNICODE` and `_UNICODE` are defined...

(and quite frankly, the definition of those two should be moved to `VC-common`, and the comment as well, but that might be for another PR)

File: Configurations/50-win-onecore.conf [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    },
+    "VC-WIN64-ARM" => {
+        inherit_from    => [ "VC-noCE-common" ],
+        defines         => add("_ARM_WINAPI_PARTITION_DESKTOP_SDK_AVAILABLE"),
+        bn_ops          => "SIXTY_FOUR_BIT RC4_CHAR EXPORT_VAR_AS_FN",
+        lflags          => add("/NODEFAULTLIB:kernel32.lib"),
+        ex_libs         => "onecore.lib onecore_downlevel.lib",
+        multilib        => "-arm64",
+    },
+);
```
Comment (levitte) [inline Configurations/50-win-onecore.conf]: You should add those config target names in [`INSTALL`](https://github.com/openssl/openssl/blob/master/INSTALL#L95)

File: Configurations/10-main.conf (hunk 6 above)
Comment (dot-asm) [reply Configurations/10-main.conf]: Let's save this for another merge request. I.e. some clean up on defines and commentary.

File: Configurations/50-win-onecore.conf (hunk 7 above)
Comment (dot-asm) [reply Configurations/50-win-onecore.conf]: If comment is about ARM targets, then I'd be against, because they are not actually supported. As commentary mentions, ARM64 is not tested *at all*, and ARM has received very limited [low-level algorithm] testing. If comment applies even to ONECORE, I'd say let's get rid of onecore_downlevel first. As it is now, it's really work in progress. On related side note one can wonder is it actually needed to list *all* VC targets in INSTALL? We don't list all Linux, BSD, Solaris, HP-UX ones :-) :-) :-)

Comment (levitte) [reply Configurations/50-win-onecore.conf]: Note that for the Unix flavored platforms, we document the use of `./config`, which automates discovery.  For Windows, we don't, and instead list the available config targets.  I see nothing weird with that...

So why don't we at least list those with expect people to start using?  From listening to the conversation on Onecore, I suspect those will become popular config targets pretty soon, so I do think those should be listed at the very least.

Comment (dot-asm) [reply Configurations/50-win-onecore.conf]: > So why don't we at least list those with expect people to start using? From listening to the conversation on Onecore, I suspect those will become popular config targets pretty soon, 

Same conversation suggests that it won't unless we drop onecore_downlevel :-) Once again, as it stands now, Onecore is work in progress, and will remain work in progress after this commit. And suggestion is to add new targets to INSTALL *after* we omit onecore_downlevel.

Comment (levitte) [reply Configurations/50-win-onecore.conf]: Okie, I'm fine with that.  Let's just not forget it.

**Comment by dot-asm** [comment]
On related side note. something to ponder over at some point. Why not
```
    "VC-ONECORE" => {
        template        => 1,
        lflags          => add("/NODEFAULTLIB:kernel32.lib"),
        ex_libs         => "onecore.lib onecore_downlevel.lib",
    "VC-WIN32-ONECORE" => {
        inherit_from    => [ "VC-WIN32", "VC-ONECORE" ],
    },
```
Problem there was that template ex_libs'  get *merged*, while what one wants is override. I.e. one wants ex_libs from VC-ONECORE to *replace* ex_libs from VC-WIN32.

**Comment by levitte** [comment]
> On related side note. something to ponder over at some point. Why not
> ```
>     "VC-ONECORE" => {
>         template        => 1,
>         lflags          => add("/NODEFAULTLIB:kernel32.lib"),
>         ex_libs         => "onecore.lib onecore_downlevel.lib",
>     },
>     "VC-WIN32-ONECORE" => {
>         inherit_from    => [ "VC-WIN32", "VC-ONECORE" ],
>     },
> ```
> Problem there was that template ex_libs'  get *merged*, while what one wants is override. I.e. one wants ex_libs from VC-ONECORE to *replace* ex_libs from VC-WIN32.

I did ponder such things when designing the `inherit_from` mechanism, and yes, it's designed to merge templates that are in the same `inherit_from` list.  The main reason is that I see the list of `inherit_from` as an unordered set, and from the point of view, overriding is impossible.
If you're thinking of adding an overriding mechanism, I would suggest choosing another attribute for it, to make it explicit.  I do note the `add` in `VC-ONECORE`, and understand the intention, but that will have to be treated with quite a lot of care, methinks.

Anyway, this seems like side discussion, maybe it should move somewhere else, so we don't clutter *this* PR with unrelated stuff?

**Comment by levitte** [comment]
> Remark did open with "something to ponder over at **some point**"

I did notice, but these kinds of points tend to be lost when talked about in unrelated PRs ;-)
